### ISSUE TITLE
no-loop-func shouldn't warn when a function is used inside a forEach or such functions

### ISSUE THREAD

**Issue by julienw** [eslint/eslint#5044]
**What version are you using?**
2.0.0-beta.1

**What did you do?**
I tested this code using the no-loop-func rule:

```javascript
  for (const score in scores) {
    const letters = scores[score];
    letters.split('').forEach(letter => {
      result[letter] = score;
    });
  }
```
Because of the function in forEach, the no-loop-func gives a warning here. Yet this is not an issue because forEach's parameter actually runs right away, it's not an asynchronous callback.

This should be relaxed at least for all Array iteration methods (also for Map and Set, but this is also a `forEach` method; maybe we can accept that all forEach/map/etc methods will behave the same).
And also the Promise constructor.
I'd be happy enough with a rule configuration.

**Comment by eslintbot** [comment]
@julienw Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
The rule is warning not because of the `forEach`, but because of the outer `for...in` loop. You are creating a new function for each iteration of the outer loop. I think the rule is behaving correctly. It will actually not flag functions inside `forEach`.

**Comment by mysticatea** [comment]
Thank you for this issue.

I know that immediate called functions are no problem even if those exist in a loop. But it's too hard to detect whether it's immediate called or not in static analysis. So I guess `no-loop-func` is warning all functions in a loop.

**Comment by julienw** [comment]
> The rule is warning not because of the forEach, but because of the outer for...in loop. You are creating a new function for each iteration of the outer loop. I think the rule is behaving correctly. It will actually not flag functions inside forEach.

Yes I know why the rule is warning. I know it's also happening because I access `result` that's out of the loop.

My proposal here is that we could assume that methods called forEach/map/reduce/etc immediately execute their function parameter. No need for complex static analysis.

This behavior would be behind a new config flag for the rule.

**Comment by michaelficarra** [comment]
How about only ignoring `[].forEach.call(..., function(){ ... })`? I could get behind that.

**Comment by ilyavolodin** [comment]
Could you explain why you think just because the function is an argument for `forEach` it's fine to use inside the loop? My understanding was that one of the reasons behind this rule was that function creating is expansive, due to creation of new scope. And doing that inside the loop would be even more costly. I'm not sure why does it matter that the function is an immediate call.

**Comment by michaelficarra** [comment]
@ilyavolodin No, it's a scoping thing. He can't create the function outside the loop because it closes over a loop variable.

```js
for (var a = [], i = 0; i < 5; ++i) {
  a.push(function(){ return i; });
}
a.map(function(f){ return f(); }); // [5, 5, 5, 5, 5]
```

But this reminds me that @julienw is using const, which is let-scooped, so wouldn't have the same problem as the var in my example. That should be made an exception.

**Comment by julienw** [comment]
What @michaelficarra said. (see also https://github.com/eslint/eslint/blob/master/docs/rules/no-loop-func.md for more explanation)

The doc above says that the code already checks for let-scopped variables. But in my example I'm using a variable that's outside of the `for` scope (`result`), that's why the rule gives a warning.

Yet with forEach and friends there is no issue. (Thinking about the performance is doing premature optimization IMO). Of course I could use `Objets.keys.forEach` instead of `for in` and then the issue would disappear by itself :) Yet I think adding a flag could be useful here.

But again I'm not suggesting to change the default behavior, because I understand some users could want this behavior and the change could be a surprise to them. Yet adding a new configuration flag for the rule would be nice.

**Comment by michaelficarra** [comment]
@julienw I don't think we'd even need a flag if we limited the exception to what I suggested. Would you be okay with that?

**Comment by julienw** [comment]
Sorry, I'm not sure I understood the rationale behind your suggestion :)

**Comment by michaelficarra** [comment]
We can be reasonably confident that, when written that way, we're using the `forEach` from the Array prototype.

**Comment by platinumazure** [comment]
@michaelficarra :-1: The utility to maintenance cost ratio for your proposal seems low. Not everyone would think to try that style of invoking forEach and it would have to be called out in docs, and it doesn't really gain all that much. On the other hand, the code to do that isn't super trivial either. I don't much like it, sorry.

**Comment by nzakas** [comment]
Creating exceptions for array methods is far too specific a change for this rule. 

It looks like we already have exceptions for using `let` or `const` loops, so the question is, why is this warning at all right now?

**Comment by julienw** [comment]
Hey Nicholas !

> Creating exceptions for array methods is far too specific a change for this rule.

My proposal is to create an exception for all methods that are called `forEach` `map` `filter` `reduce` `reduceRight`, `find`, findIndex`, `some` , `every`, `sort`. The rationale is that all methods call like this will likely have the same behaviors than Array instance methods, I mean the callback will be immediately called.

> It looks like we already have exceptions for using let or const loops, so the question is, why is this warning at all right now?

I said it earlier, my code actually use a var from outside the loop (`result`).

**Comment by julienw** [comment]
> We can be reasonably confident that, when written that way, we're using the forEach from the Array prototype.

I agree with @platinumazure here, I don't think this is useful for anybody. I wouldn't write unreadable code only to accommodate an eslint rule :)

**Comment by michaelficarra** [comment]
@julienw Well we can't assume that any method named `map` will behave like `[].map`, since Stream libraries like highland.js exist and I'm sure there's a Futures/Promises library that defines a `map` method.

**Comment by julienw** [comment]
yeah maybe it's an assumption we can't make.

**Comment by mysticatea** [comment]
BTW, in fact, is the `result` variable not modified after the loop?
If so, this pattern is safe. We can confirm whether the function is safe or not by checking whether or not `WriteReference` exists after the loop, I guess.

**Comment by julienw** [comment]
In my case it's not but it's returned from the function, so I think it would be very difficult to find out it's not modified.

So after some sleep, I understand that we can't whitelist forEach-etc methods because it would be too error-prone.

But I also understand that the rule is trying to check for 2 different patterns (according to http://jslinterrors.com/dont-make-functions-within-a-loop/):

1. The dangerous pattern of capturing an unscoped variable in the closure. We all agree that this should be checked, but my code above does not have such issues.

2. The pattern of creating a function in a loop which could have performance issues. Here I think it's a premature optimization: this could actually be a problem for big loops. And actually for this pattern we shouldn't whitelist const/let scoped variables either.

So I'd suggest to actually split this rule to check these patterns in 2 different rules. What do you think ?

If everyone agrees we can change the title of this bug which does not reflect what is needed anymore ;)

**Comment by nzakas** [comment]
Agree with @mysticatea, that seems like the correct approach. 

@julienw we don't have rules related to performance. Performance is always based on the engine, and as such, performance characteristics change as the engine changes.

**Comment by julienw** [comment]
@nzakas, I agree with this, but that's not how the test is behaving :)

**Comment by nzakas** [comment]
@julienw what does "this" refer to in your statement?

**Comment by julienw** [comment]
@nzakas I agree with the fact that rules in eslint should not be related to performance. Yet this rule seems to check this because it warns (not only, but also) about code that's correct but could have performance problem.

Now about @mysticatea approach. What about this code:
```javascript
// scores is an object coming as a parameter
// its values are strings.
let result = {};
for (const score in scores) {
  const letters = scores[score];
  letters.split('').forEach(letter => {
    result[letter] = score;
  });
}
result.__default = 6;
```

This would fail the test, but yet it's safe.

But we know it's safe only because we know forEach is immediately invoking its parameter. So I understand why it's difficult, and if we don't want to handle this case it should be documented.

**Comment by mysticatea** [comment]
`result.__default = 6;` makes a `ReadReference` of `result`. So it would not be warned.
Currently, eslint is not handling property references.

**Comment by julienw** [comment]
Mmm ok got it. Then I agree with your approach !

**Comment by nzakas** [comment]
Just to be clear for everyone: this rule makes no consideration based on performance. We are purely doing static analysis for possible errors. 

@mysticatea do we need a change here?

**Comment by mysticatea** [comment]
@nzakas my idea on this issue is that no-loop-func doesn't need to warn a function which isn't using modified variables of outer scopes. It would reduce false positive, I thought.

**Comment by nzakas** [comment]
Got it, sounds good.

**Comment by s7dhansh** [comment]
Same issue. In my case the variable is modified before and after the loop. Still the code is perfectly safe. 

eg. 	
```
for .... { details[i].map(function(it) { summary.addToSet(it); }); }
for .... { specifications[i].map(function(it) { summary.addToSet(it); }); }
```

Are there no exceptions that we can define for no-loop-func?

**Comment by dandv** [comment]
```js
let space = MAX_SPACE;
while (space > DANGER) {
    const capacities = await Promise.all([...]);  // array of signed integers
    capacities.forEach(x => space += x);
}
```

What am I *actually* doing wrong that I get `no-return-assign` and `no-loop-func`?

**Comment by krassowski** [comment]
Similar issue, using d3.js framework which uses callbacks heavily (to execute them immediately, in order to perform DOM modifications). My example:

```js
var width_per_group = 10
var width = groups.length * width_per_group
var shift = - width / 2

for(var i = 0; i < groups.length; i++)
{
    groups[i].attr('transform', function(node) {return do_some_stuff(node, shift)})
    shift += width_per_group
}
```

I think that this is quite common pattern. Please notice, that the callback is executed directly in the loop and it is no longer stored anywhere (what can be indeed difficult to determine using static analysis).

Do I understand correctly that 'no-loop-func' resulting from linting the code above is a false positive?

**Comment by not-an-aardvark** [comment]
@krassowski No, it's not a false positive. You should move the function outside of the loop:

```js
var width_per_group = 10
var width = groups.length * width_per_group
var shift = - width / 2

function handleNode(node) {
    return do_some_stuff(node, shift);
}

for(var i = 0; i < groups.length; i++)
{
    groups[i].attr('transform', handleNode)
    shift += width_per_group
}
```

**Comment by krassowski** [comment]
@not-an-aardvark Thank you, your answer shows a better idea for implementing the desired behavior.

But still, was my snippet really incorrect (not speaking about more work for interpreter due to additional function declarations) or bug prone?
